src/android/java/io/jxcore/node/BtConnectorHelper.java

```diff
@@ -191,122 +155,141 @@ public String  MakeGUID() {
 
     public void BeginConnectPeer(String toPeerId, ConnectStatusCallback connectStatusCallback) {
         ConnectStatusCallback tmpCallback = connectStatusCallback;
-
-        if(tmpCallback == null) {
+        if (tmpCallback == null) {
             //nothing we should do, since we can not update progress
             print_debug("BeginConnectPeer callback is NULL !!!!!!");
-        }else if(mBtToRequestSocket != null) {
-            tmpCallback.ConnectionStatusUpdate("Already connected to " + mBtToRequestSocket.GetPeerId(),-1);
-        }else {
-            ServiceItem selectedDevice = null;
-            if (lastAvailableList != null) {
-                for (int i = 0; i < lastAvailableList.size(); i++) {
-                    if (lastAvailableList.get(i).peerId.contentEquals(toPeerId)) {
-                        selectedDevice = lastAvailableList.get(i);
-                        break;
-                    }
-                }
-            }
+            return;
+        }
 
-            if (selectedDevice != null) {
-                if (mBTConnector != null) {
-                    BTConnector.TryConnectReturnValues retVal = mBTConnector.TryConnect(selectedDevice);
-                    if (retVal == BTConnector.TryConnectReturnValues.Connecting) {
-                        //all is ok, lets wait callbacks, and for them lets copy the callback here
-                        mConnectStatusCallback = connectStatusCallback;
-                    } else if (retVal == BTConnector.TryConnectReturnValues.NoSelectedDevice) {
-                        // we do check this already, thus we should not get this ever.
-                        tmpCallback.ConnectionStatusUpdate("Device Address for " + toPeerId + " not found from Discovered device list.",-1);
-                    } else if (retVal == BTConnector.TryConnectReturnValues.AlreadyAttemptingToConnect) {
-                        tmpCallback.ConnectionStatusUpdate("There is already one connection attempt progressing.",-1);
-                    } else if (retVal == BTConnector.TryConnectReturnValues.BTDeviceFetchFailed) {
-                        tmpCallback.ConnectionStatusUpdate("Bluetooth API failed to get Bluetooth device for the address : " + selectedDevice.peerAddress,-1);
-                    }
+        BtToRequestSocket tmpToReqSoc = mBtToRequestSocket;
+        if (tmpToReqSoc != null) {
+            tmpCallback.ConnectionStatusUpdate("Already connected to " + tmpToReqSoc.GetPeerId(), -1);
+            return;
+        }
 
-                } else {
-                    tmpCallback.ConnectionStatusUpdate("Device conenctivity not started, please call StartBroadcasting before attempting to connnect",-1);
+        ServiceItem selectedDevice = null;
+        if (lastAvailableList != null) {
+            for (int i = 0; i < lastAvailableList.size(); i++) {
+                if (lastAvailableList.get(i).peerId.contentEquals(toPeerId)) {
```

Comment:
Might it make sense to make lastAvailableList into something like a ConcurrentHashMap? No need to implement your own search function. Entries could be added based on peerId.